src/node_i18n.cc

```diff
@@ -185,13 +211,338 @@ static void ToASCII(const FunctionCallbackInfo<Value>& args) {
+  ucnv_convertEx(to.conv, from.conv,
+                 &target, target + limit,
+                 &data, ts_obj_data + ts_obj_length,
+                 NULL, NULL, NULL, NULL,
+                 true, true,
+                 &status);
+
+  if (U_SUCCESS(status)) {
+    len = target - *buf;
+    args.GetReturnValue().Set(
+      Buffer::Copy(env->isolate(), *buf, len).ToLocalChecked());
+  } else {
+    env->ThrowError(u_errorName(status));
+  }
+}
+
+// Converts to UCS2 from ISO-8859-1 and US-ASCII
+// args[0] is the encoding to convert from
+// args[1] must be a buffer instance
+void ConvertToUcs2(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsString());
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[1]);
+
+  Utf8Value name(env->isolate(), args[0]);
+
+  SPREAD_ARG(args[1], ts_obj);
+
+  UErrorCode status = U_ZERO_ERROR;
+  size_t len;
+
+  MaybeStackBuffer<UChar> buf(ts_obj_length);
+
+  Converter from(env, *name);
+  if (from.conv == nullptr)
+    return;
+
+  len = ucnv_toUChars(from.conv,
+                      *buf, ts_obj_length << 1,
+                      ts_obj_data, ts_obj_length,
+                      &status);
+
+  if (U_SUCCESS(status)) {
+    if (IsBigEndian()) {
+      uint16_t* dst = reinterpret_cast<uint16_t*>(*buf);
+      SwapBytes(dst, dst, len);
+    }
+    args.GetReturnValue().Set(
+      Buffer::Copy(env->isolate(),
+                  reinterpret_cast<char*>(*buf),
+                  ts_obj_length << 1).ToLocalChecked());
+  } else {
+    env->ThrowError(u_errorName(status));
+  }
+}
+
+// Convert to a named encoding from UCS2
+// args[0] is the name of the encoding being converted to
+// args[1] must be a buffer instance
+void ConvertFromUcs2(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsString());
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[1]);
+
+  Utf8Value name(env->isolate(), args[0]);
+  SPREAD_ARG(args[1], ts_obj);
+
+  UErrorCode status = U_ZERO_ERROR;
+
+  UChar* source = nullptr;
+  MaybeStackBuffer<UChar> swapspace;
+  const size_t length = ts_obj_length >> 1;
+
+  if (IsLittleEndian()) {
+    source = reinterpret_cast<UChar*>(ts_obj_data);
+  } else {
+    swapspace.AllocateSufficientStorage(length);
+    for (size_t n = 0, i = 0; i < length; n += 2, i += 1) {
+      const uint8_t hi = static_cast<uint8_t>(ts_obj_data[n + 0]);
+      const uint8_t lo = static_cast<uint8_t>(ts_obj_data[n + 1]);
+      swapspace[i] = (hi << 8) | lo;
+    }
+    source = *swapspace;
+  }
+
+  uint32_t len;
+  MaybeStackBuffer<char> buf(length);
+
+  Converter to(env, *name);
+  if (to.conv == nullptr)
+    return;
+  ucnv_setSubstChars(to.conv, "?", 1, &status);
+  status = U_ZERO_ERROR;
+
+  len = ucnv_fromUChars(to.conv, *buf, length, source, length, &status);
+
+  if (U_SUCCESS(status)) {
+    args.GetReturnValue().Set(
+      Buffer::Copy(env->isolate(), *buf, len).ToLocalChecked());
```

Comment:
there a way to grab the pointer on `buf` if the allocation was larger than `kStackStorageSize` instead of doing a `memcpy()` every time?